From: Rod Whitby <rod@whitby.id.au>
Date: Sat, 10 Feb 2007 09:26:48 +0000 (+0000)
Subject: [MTD] Fix RedBoot partition parsing regression harder.
X-Git-Tag: v2.6.21-rc1^0~66^2~12
X-Git-Url: http://www2.kernel.org/git/?p=linux%2Fkernel%2Fgit%2Ftorvalds%2Flinux-2.6.git;a=commitdiff_plain;h=11192146e57bc8e58865e7d9c2497f66a4f7e6e7

[MTD] Fix RedBoot partition parsing regression harder.

Correct the location of the recalculation of the FIS directory size,
and also add the same recalculation for the byte-swapped case.

Signed-off-by: Rod Whitby <rod@whitby.id.au>
Signed-off-by: David Woodhouse <dwmw2@infradead.org>
---

diff --git a/drivers/mtd/redboot.c b/drivers/mtd/redboot.c
index 5a31f49..a459ffa 100644
--- a/drivers/mtd/redboot.c
+++ b/drivers/mtd/redboot.c
@@ -96,6 +96,8 @@ static int parse_redboot_partitions(struct mtd_info *master,
 			 */
 			if (swab32(buf[i].size) == master->erasesize) {
 				int j;
+				/* Update numslots based on actual FIS directory size */
+				numslots = swab32(buf[i].size) / sizeof (struct fis_image_desc);
 				for (j = 0; j < numslots; ++j) {
 
 					/* A single 0xff denotes a deleted entry.
@@ -120,11 +122,11 @@ static int parse_redboot_partitions(struct mtd_info *master,
 					swab32s(&buf[j].desc_cksum);
 					swab32s(&buf[j].file_cksum);
 				}
+			} else {
+				/* Update numslots based on actual FIS directory size */
+				numslots = buf[i].size / sizeof(struct fis_image_desc);
 			}
 			break;
-		} else if (buf[i].size != -1) {
-			/* re-calculate of real numslots */
-			numslots = buf[i].size / sizeof(struct fis_image_desc);
 		}
 	}
 	if (i == numslots) {

